_posts/2023-06-23-run-and-debug-rust-verification-harnesses-natively-on-vscode.md

```diff
@@ -0,0 +1,179 @@
+
+#### Generate a counterexample unit test
+
+Next, we’ll generate the unit test by clicking on the `Run Concrete Playback for stretched_rectangle_can_hold_original` link that appears through a blue link on the error banner.
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/vs-code-images/generate-counter-example.gif" style="box-shadow: 0 2px 4px rgba(0, 0, 0, 0.5);" alt="generate-unit-test" />
+
+By simply clicking on a link, we have our counterexample unit test pasted directly below the harness.
+This is what the unit test looks like:
+
+```rust
+#[test]
+fn kani_concrete_playback_stretched_rectangle_can_hold_original() {
+    let concrete_vals: Vec<Vec<u8>> = vec![
+        // 4611686018427387904ul
+        vec![0, 0, 0, 0, 0, 0, 0, 64],
+        // 0ul
+        vec![0, 0, 0, 0, 0, 0, 0, 0],
+        // 2ul
+        vec![2, 0, 0, 0, 0, 0, 0, 0],
+    ];
+    kani::concrete_playback_run(concrete_vals, stretched_rectangle_can_hold_original);
+}
+```
+
+You can see in the GIF above that the source is now annotated with two options on top of the generated unit test, `Run Test (Kani)` and `Debug Test (Kani)`, which allow you to run and debug the test just like any other Rust unit test.
+
+#### Run Kani-generated test
+
+Running the unit test using the `Run Test (Kani)` button shows us what we were expecting -- that the unit test is failing.
+This is because the unit test is using the counterexample to invoke the function `stretched_rectangle_can_hold_original`.
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/vs-code-images/run-concrete-playback-test.png" alt="run-concrete-playback-test" />
+
+### Debug Kani unit test
+
+In order to peek under the hood to find out the missing assumptions that lead to unexpected behavior, it is really important to look at the concrete counterexamples for which our assertions fail.
+By setting breakpoints and clicking the `Debug Test (Kani)` button, you are taken into the debugger which allows you to inspect the specific values for which the assertion fails.
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/vs-code-images/run-debugger.gif" style="box-shadow: 0 2px 4px rgba(0, 0, 0, 0.5);" alt="run-debugger" />
+
+In our case, we can see that for `original.height = 0` , the larger rectangle’s height or `larger.height` also stays 0, which shows that for that counterexample, the property `can_hold` does not hold.
+
+### And finally, verify the harness with the right assumptions
+
+Now that we know that for `original.width = 0`, our assertion fails, we can repeat the experiment with explicit assumptions.
+The experiments should reveal that for all parameters, having a 0 value will cause the assertion to fail.
+Additionally, there is a problem if `factor` is `1` because in this case stretch will return `Some(...)` but the stretched rectangle will be the same size as the original.
```

Comment:
```suggestion
Additionally, there is a problem if `factor` is `1` because in this case `stretch` will return `Some(...)`, but the stretched rectangle will be the same size as the original.
```